### PR TITLE
makefile, test and scripts updates to use go toolchain instead of bazel.

### PR DESCRIPTION
**Pull request by costinm**
Multiple changes to improve build and testing with the go toolchain. 

- presubmit changes to use go tools.

- stop using the k8s config linked from the user home to some dir that gets picked by bazel.
It is pretty dangerous and flaky - unit tests are actually run against whatever cluster the 
developer has set, may affect a live cluster. It also pollutes the source tree and is not standard
go. Refactored the files using config to use a common method.

- script equivalent with the circleci local unit test env, consisting of etcd+apiserver.

- checkin few missing files, pending Dougz full move to make/shell generation for automation.
Current circleci build simply excludes the tests.

```release-note
- istio builds using go toolchain instead of bazel. 
```

### CODE REVIEW

File: bin/testEnvLocalK8S.sh [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+# expect istio scripts to be under $GOPATH/src/istio.io/istio/bin/...
+export TOP=$(cd $(dirname $0)/../../../..; pwd)
+
+export GOPATH=${TOP}
+export PATH=${GOPATH}/bin:${PATH}
+export OUT=${TOP}/out
+
+# components used in the test (starting with circleci for consistency, eventually ci will use this)
+export K8S_VER=v1.7.4
```
Comment (kyessenov) [inline bin/testEnvLocalK8S.sh]: This is very old. 1.9 just got released.

File: broker/pkg/platform/kube/crd/client_test.go [hunk 2]
```diff
@@ -26,23 +24,11 @@ import (
...
-
-	// For Bazel sandbox, the config is mounted in a different location
-	// as specified in prow/broker-presubmit.sh
-	if _, err = os.Stat(f); err != nil {
-		wd, _ := os.Getwd()
-		f = wd + "/../config"
-	}
-
-	return f
+	return k8s.Kubeconfig("/../../../../../pilot/platform/kube/config")
```
Comment (kyessenov) [inline broker/pkg/platform/kube/crd/client_test.go]: Where's the file being symlinked?

File: mixer/tools/codegen/pkg/interfacegen/testdata/apa_tmpl.pb.go [hunk 3]
```diff
@@ -1,11 +1,11 @@
 // Code generated by protoc-gen-gogo. DO NOT EDIT.
-// source: bazel-out/local-fastbuild/genfiles/mixer/tools/codegen/pkg/interfacegen/testdata/apa_tmpl.proto
+// source: bazel-out/k8-fastbuild/genfiles/mixer/tools/codegen/pkg/interfacegen/testdata/apa_tmpl.proto
```
Comment (kyessenov) [inline mixer/tools/codegen/pkg/interfacegen/testdata/apa_tmpl.pb.go]: Unnecessary change?

File: bin/testEnvLocalK8S.sh (hunk 1 above)
Comment (costinm) [reply bin/testEnvLocalK8S.sh]: Using same version that circleci is using, and I believe it's the min. version of k8s we support.

We can run test envs with 1.9 as well.

File: broker/pkg/platform/kube/crd/client_test.go (hunk 2 above)
Comment (costinm) [reply broker/pkg/platform/kube/crd/client_test.go]: In the old bazel - I'm not changing anything, I assume scripts will do it.

For go/make: it is no longer used. Once bazel is out I'll also remove this line, and only use 
KUBECONFIG and the localhost/standalone apiserver conf for unit tests, and whatever the new
testing framework uses for integration ( most tests can be run in multiple environments ).

File: mixer/tools/codegen/pkg/interfacegen/testdata/apa_tmpl.pb.go (hunk 3 above)
Comment (costinm) [reply mixer/tools/codegen/pkg/interfacegen/testdata/apa_tmpl.pb.go]: Automated scripts... Will try to remove it, but once Doug has his PR in I assume all those will
be replaced.

**Comment by kyessenov** [comment]
Looks reasonable. I'd like to keep a check on the bash scripting that will be happening with the Makefile approach. I think what you have here for spinning up local apiserver is OK, but if there's a third party project that can do that (even if it's minikube), I'd prefer that.

Also, I see a lot of `go install`. Should we be using `go build -i` instead? That way we don't mess with other projects in the same $GOROOT (which I'm sure many developers have).

**Comment by costinm** [comment]
On the shell script: it's replicating what circleci has been doing for now, but eventually this will be migrated  into the new test framework, as one of the environments.  

For 'unit test' having the etcd/apiserver is a reasonable option, but the goal is to be able to run
one test in multiple environments - however this PR is only focusing on the short term 
migration to go and consistency with the ci.

File: Makefile [hunk 4]
```diff
@@ -123,6 +148,93 @@ check.lint: ; $(info $(H) running golint on packages...)
...
+
+.PHONY: mixs
+mixs: vendor
+	go install istio.io/istio/mixer/cmd/mixs
+
+.PHONY: mixc
+mixc: vendor
+	go install istio.io/istio/mixer/cmd/mixs
+
+go-build: pilot istioctl pilot-agent sidecar-initializer mixs mixc
```
Comment (jwendell) [inline Makefile]: I wonder if it should depend also on `depend.vendor` or similar. Or it's up to the user to deal with dependencies? Current build (with bazel) does that automatically, correct?

File: bin/testEnvLocalK8S.sh [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    ${TOP}/bin/etcd > ${LOG_DIR}/etcd.log &
+    echo $! > $LOG_DIR/etcd.pid
+
+    ${TOP}/bin/kube-apiserver --etcd-servers http://127.0.0.1:2379 \
+        --client-ca-file ${CERTDIR}/k8sca.crt \
+        --requestheader-client-ca-file ${CERTDIR}/k8sca.crt \
+        --tls-cert-file ${CERTDIR}/apiserver.crt \
+        --tls-private-key-file ${CERTDIR}/apiserver.key \
+        --service-cluster-ip-range 10.99.0.0/16 \
+        --port 8080 -v 2 --insecure-bind-address 0.0.0.0 \
```
[the remark this answers is not in the document]
Comment (costinm) [reply bin/testEnvLocalK8S.sh]: The port - yes, I'm planning to do this - current code is cut&pasted from circleci, but for running on 
local machine it probably needs more customization. However not top priority, after it works we can
fine tune it... (and better yet - this can be moved into the new test framework !)

File: envsetup.sh [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+export TOP=`pwd`
+
+# Used in the shell scripts.
+export ISTIO_SRC=$TOP
+export GOPATH=$ISTIO_SRC/go
+export ISTIO_GO=$GOPATH/src/istio.io/istio
+
+# hub used to push images.
+export HUB=${ISTIO_HUB:-grc.io/istio-testing}
```
Comment (rkpagadala) [inline envsetup.sh]: typo gcr.io/istio-testing

File: Makefile [hunk 7]
```diff
@@ -29,13 +44,19 @@ BAZEL_STARTUP_ARGS ?=
...
+
 hub = ""
 tag = ""
 
 ifneq ($(strip $(HUB)),)
 	hub =-hub ${HUB}
 endif
 
+# If tag not explicitly set in users' .istiorc or command line, default to the git sha.
+TAG ?= $(shell git rev-parse --verify HEAD)
```
Comment (jwendell) [inline Makefile]: This won't work when users download a tarball of a released version. Should the fallback be the version number (like having a VERSION file)?

File: Makefile [hunk 8]
```diff
@@ -18,8 +18,23 @@
 TOP := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
 SHELL := /bin/bash
 
+# Make sure GOPATH is set based on the executing Makefile and workspace. Will override
+# GOPATH from the env.
+export GOPATH= $(shell cd ../../..; pwd)
```
Comment (jwendell) [inline Makefile]: I'm curious why we need to override the env variable

**Comment by jwendell** [comment]
@kyessenov @costinm Instead of changing `go install` with `go build -i`, what about just `go build`? `-i` flag doesn't play well with system-wide golang installations as it tries to install things in system-wide dirs. Also, in 1.10 `go build` will cache artifacts automatically, making `-i` usage for speed purposes useless...

File: Makefile [hunk 9]
```diff
@@ -78,12 +99,22 @@ depend.status: Gopkg.lock ; $(info $(H) reporting dependencies status...)
...
 	$(Q) dep ensure
 
 depend.graph: Gopkg.lock ; $(info $(H) visualizing dependency graph...)
 	$(Q) dep status -dot | dot -T png | display
 
+# Re-create the vendor directory, if it doesn't exist, using the checked in lock file
+depend.vendor: vendor
+	$(Q) dep ensure -vendor-only
+
+vendor:
```
Comment (jwendell) [inline Makefile]: Does this need a .PHONY too?

File: .circleci/config.yml [hunk 10]
```diff
@@ -233,11 +233,34 @@ jobs:
...
+      - checkout
+      - restore_cache:
+          keys:
+            - dep-cache-{{ checksum "Gopkg.toml" }}-{{ checksum "WORKSPACE" }}
+      - run:
+          command: |
+            cd /go/src/istio.io/istio
+            if [ ! -d vendor ]; then
+               dep ensure
+            fi
```
Comment (jwendell) [inline .circleci/config.yml]: What about just call `make depend`?

File: Makefile [hunk 11]
```diff
@@ -123,6 +154,96 @@ check.lint: ; $(info $(H) running golint on packages...)
...
+.PHONY: mixs
+mixs: vendor
+	go install istio.io/istio/mixer/cmd/mixs
+
+.PHONY: mixc
+mixc: vendor
+	go install istio.io/istio/mixer/cmd/mixs
+
+go-build: pilot istioctl pilot-agent sidecar-initializer mixs mixc
+
```
Comment (jwendell) [inline Makefile]: It depends on `vendor` target, which calls `dep ensure -update` which does not work on a fresh install, because there's no `vendor` dir yet. So, for fresh installs we should call `dep ensure`. Therefore we could either modify the `vendor` target or instead depend on `depend` target.

File: Makefile (hunk 7 above)
Comment (costinm) [reply Makefile]: That's what the presubmit is doing - we can change it later, but as is TAG can be set by user either 
via command line or env variable - this is only the default.

File: Makefile (hunk 8 above)
Comment (costinm) [reply Makefile]: Concerns about externally set wrong GOPATH (maybe including other things )

File: Makefile (hunk 9 above)
Comment (costinm) [reply Makefile]: Vendor is not phony I think - if the dir is missing we need to run dep ensure. 

But likely there are missing tags in the .phony, will go over the file again after I get all
the tests passing ( and probably in separate PRs)

File: Makefile (hunk 11 above)
Comment (costinm) [reply Makefile]: Got it - so far it seems it worked by accident :-), circleci is doing dep ensure before so vendor
exists. Will fix.

File: Makefile [hunk 12]
```diff
@@ -123,6 +154,96 @@ check.lint: ; $(info $(H) running golint on packages...)
...
+sidecar-initializer: vendor
+	go install istio.io/istio/pilot/cmd/sidecar-initializer
+
+.PHONY: mixs
+mixs: vendor
+	go install istio.io/istio/mixer/cmd/mixs
+
+.PHONY: mixc
+mixc: vendor
+	go install istio.io/istio/mixer/cmd/mixs
```
Comment (jwendell) [inline Makefile]: Typo: mixc

**Comment by jwendell** [comment]
There's no target for building istio_ca (security). Is that on purpose?

File: Makefile [hunk 13]
```diff
@@ -120,19 +142,115 @@ check.lint: ; $(info $(H) running golint on packages...)
...
+.PHONY: istioctl
+istioctl: vendor
+	go install istio.io/istio/pilot/cmd/istioctl
+
+.PHONY: sidecar-initializer
+sidecar-initializer: vendor
+	go install istio.io/istio/pilot/cmd/sidecar-initializer
+
+.PHONY: mixs
+mixs: vendor
```
Comment (douglas-reid) [inline Makefile]: i'm not clear where this belongs, but at some point we need to add a `generate` target that can call `go generate istio.io/istio/...` before it does a full build.

File: bin/pre-commit [hunk 14]
```diff
@@ -34,8 +37,8 @@ BRANCH_NAME=$(git branch | grep '*' | sed 's/* //')
 if [ $BRANCH_NAME != '(no branch)' ]; then
     cd $ROOT
 
-    echo "updating autogen files..."
-    bin/regenerate_files.py
+    #echo "updating autogen files..."
+    #bin/regenerate_files.py
```
Comment (douglas-reid) [inline bin/pre-commit]: as above, we need something to do proper generation somewhere. i don't have strong informed opinions on where, however.

File: envsetup.sh [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+# overriding or changing the main ~/.kube/config
+#
+# For each env, create a file under $HOME/.istio/ENV_NAME
+#
+# Will set TEST_ENV and KUBECONFIG environment variables.
+#
+# Predefined:
+# - minikube: will start a regular minikube in a VM
+#
+function lunch() {
```
Comment (douglas-reid) [inline envsetup.sh]: was this written around midday :stuck_out_tongue_winking_eye: ?

s/lunch/launch (here and in the comment).

File: mixer/adapter/opa/opa_integration_test.go [hunk 16]
```diff
@@ -177,7 +177,8 @@ spec:
 `
 )
 
-func TestServer(t *testing.T) {
+//https://github.com/istio/istio/issues/2300
+func xTestServer(t *testing.T) {
```
Comment (douglas-reid) [inline mixer/adapter/opa/opa_integration_test.go]: there is an outstanding PR to fix this. Should we merge that first and then drop this change?

File: bin/pre-commit (hunk 14 above)
Comment (costinm) [reply bin/pre-commit]: I don't know if people are more likely to install the pre-submit hook (I forget sometimes), or to 
run make / test. 

But it's clear make will need to have the target - so maybe we can start with having it in the makefile.

File: envsetup.sh (hunk 15 above)
Comment (costinm) [reply envsetup.sh]: No, 8 years of android ( envsetup and lunch are the first 2 commands android devs type every day). 
https://source.android.com/setup/building

File: mixer/adapter/opa/opa_integration_test.go (hunk 16 above)
Comment (costinm) [reply mixer/adapter/opa/opa_integration_test.go]: Either way - if it gets merged before I finish (likely), I'll drop.

**Comment by costinm** [comment]
And finally prow is happy !!!

Still missing (separate PRs):
- integrate Doug's generate scripts
- deb/rpm builds using .fpm
- lots of cleanup and consistency improvements !
